File: skellig-connection-http/src/main/java/org/skellig/connection/http/model/HttpRequestDetails.java
package org.skellig.connection.http.model;

import java.util.Collections;
import java.util.Map;
import java.util.Optional;

public class HttpRequestDetails {

    private final HttpMethodName verb;
    private final String url;
    private final Map<String, String> formParams;
    private final String body;
    private final Map<String, String> headers;
    private final Map<String, String> queryParams;
    private final String username;
    private final String password;

    private HttpRequestDetails(HttpMethodName verb, String url, Map<String, String> headers,
                               Map<String, String> formParams, String body, Map<String, String> queryParams,
                               String username, String password) {
        this.verb = verb;
        this.url = url;
        this.headers = headers;
        this.formParams = formParams;
        this.queryParams = queryParams;
        this.body = body;
        this.username = username;
        this.password = password;
    }

    public HttpMethodName getVerb() {
        return verb;
    }

    public String getUrl() {
        return url;
    }

    public Map<String, String> getHeaders() {
        return headers;
    }

    public Map<String, String> getFormParams() {
        return formParams;
    }

    public String getBody() {
        return body;
    }

    public Map<String, String> getQueryParams() {
        return queryParams;
    }

    public Optional<String> getUsername() {
        return Optional.ofNullable(username);
    }

    public Optional<String> getPassword() {
        return Optional.ofNullable(password);
    }

    @Override
    public String toString() {
        return url;
    }

    public static class Builder {

        private HttpMethodName verb;
        private String url;
        private Map<String, String> headers = Collections.emptyMap();
        private Map<String, String> formParams = Collections.emptyMap();
        private Map<String, String> queryParams= Collections.emptyMap();
        private String body;
        private String username;
        private String password;

        public Builder(HttpMethodName verb) {
            this.verb = verb;
        }

        public Builder withUrl(String url) {
            this.url = url;
            return this;
        }

        public Builder withHeaders(Map<String, String> headers) {
            if (headers != null) {
                this.headers = headers;
            }
            return this;
        }

        public Builder withFormParam(Map<String, String> formParams) {
            if (formParams != null) {
                this.formParams = formParams;
            }
            return this;
        }

        public Builder withQueryParam(Map<String, String> queryParams) {
            if (queryParams != null) {
                this.queryParams = queryParams;
            }
            return this;
        }

        public Builder withBody(String body) {
            this.body = body;
            return this;
        }

        public Builder withUsername(String username) {
            this.username = username;
            return this;
        }

        public Builder withPassword(String password) {
            this.password = password;
            return this;
        }

        public HttpRequestDetails build() {
            return new HttpRequestDetails(verb, url, headers, formParams, body, queryParams, username, password);
        }
    }

}

File: skellig-test-step-processing-http/src/main/kotlin/org/skellig/teststep/processor/http/model/HttpRequestDetails.kt
package org.skellig.teststep.processor.http.model

/**
 * Class representing the details of an HTTP request.
 *
 * @property verb The HTTP method name.
 * @property url The URL of the request.
 * @property headers The headers of the request.
 * @property formParams The form parameters of the request.
 * @property body The body of the request.
 * @property queryParams The query parameters of the request.
 * @property username The username for authentication.
 * @property password The password for authentication.
 * @property timeout The timeout for the request.
 */
class HttpRequestDetails(val verb: HttpMethodName,
                         val url: String,
                         val headers: Map<String, String?>,
                         val formParams: Map<String, String?>,
                         val body: String?,
                         val queryParams: Map<String, String?>,
                         val username: String?,
                         val password: String?,
                         val timeout: Int) {

    override fun toString(): String {
        return "$verb:$url" +
                (if (headers.isNotEmpty()) ("\nheaders: $headers") else "") +
                (if (formParams.isNotEmpty()) ("\nformParams: $formParams") else "") +
                (if (queryParams.isNotEmpty()) ("\nqueryParams: $queryParams") else "") +
                (body?.let { ("\nbody: $body") } ?: "") +
                (username?.let { ("\nusername: $username") } ?: "") +
                (password?.let { ("\npassword: $password") } ?: "")
    }

    class Builder(private val verb: HttpMethodName) {

        private var url: String? = null
        private var headers = emptyMap<String, String?>()
        private var formParams = emptyMap<String, String?>()
        private var queryParams = emptyMap<String, String?>()
        private var body: String? = null
        private var username: String? = null
        private var password: String? = null
        private var timeout = 0

        fun withUrl(url: String?) = apply {
            this.url = url
        }

        fun withHeaders(headers: Map<String, String?>?) = apply {
            if (headers != null) {
                this.headers = headers
            }
        }

        fun withFormParam(formParams: Map<String, String?>?) = apply {
            if (formParams != null) {
                this.formParams = formParams
            }
        }

        fun withQueryParam(queryParams: Map<String, String?>?) = apply {
            if (queryParams != null) {
                this.queryParams = queryParams
            }
        }

        fun withBody(body: String?) = apply {
            this.body = body
        }

        fun withUsername(username: String?) = apply {
            this.username = username
        }

        fun withPassword(password: String?) = apply {
            this.password = password
        }

        fun withTimeout(timeout: Int) = apply {
            this.timeout = timeout
        }

        fun build(): HttpRequestDetails {
            return HttpRequestDetails(verb, url!!, headers, formParams, body, queryParams, username, password, timeout)
        }
    }

}